### ISSUE TITLE
Deadlock stealing a `resumed` task

### ISSUE THREAD

**Issue by gjoseph92** [dask/distributed#6159]
* Worker starts fetching key `abcd` from a peer
* The fetch gets cancelled
* Scheduler now asks the worker to _compute_ the key `abcd`
* Deadlock

Here's the (annotated) worker story for the key in question:
```yaml
# Initially we hear about the key as a dependency to fetch
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - ensure-task-exists
  - released
  - compute-task-1650391863.1634912
  - 2022-04-19 11:11:03.369210
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - released
  - fetch
  - fetch
  - {}
  - compute-task-1650391863.1634912
  - 2022-04-19 11:11:03.369474
- - gather-dependencies
  - tls://10.0.13.152:38353
  - - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 8))
    - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - ensure-communicating-1650391867.838159
  - 2022-04-19 11:11:07.838874
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - fetch
  - flight
  - flight
  - {}
  - ensure-communicating-1650391867.838159
  - 2022-04-19 11:11:07.838897
# We go try to fetch it. I think we're talking to a worker that's still alive,
# but so locked up under memory pressure it never responds—see https://github.com/dask/distributed/issues/6110
- - request-dep
  - tls://10.0.13.152:38353
  - - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 8))
    - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - ensure-communicating-1650391867.838159
  - 2022-04-19 11:11:07.848737
# This is really weird. After the stuck worker's TTL expires (https://github.com/dask/distributed/issues/6110#issuecomment-1102959742),
# the scheduler removes it. But the log we're looking at _isn't from the stuck worker_.
# So why is `Scheduler.transition_processing_released` sending a `free-keys` to this worker?
# That would indicate the scheduler's `processing_on` pointed to this worker.
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - flight
  - released
  - cancelled
  - {}
  - processing-released-1650392050.5300016
  - 2022-04-19 11:14:10.571274
# Regardless, the scheduler then asks us to compute (not fetch!) this task.
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - compute-task
  - compute-task-1650392058.1483586
  - 2022-04-19 11:14:18.165346
# We already know about it---the fetch was just cancelled by `processing-released`---so it stays in cancelled,
# with a recommendation to `(resumed, waiting)`
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - cancelled
  - waiting
  - cancelled
  - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1)):
    - resumed
    - waiting
  - compute-task-1650392058.1483586
  - 2022-04-19 11:14:18.165884
# It goes into resumed and nothing ever happens again
- - ('split-shuffle-1-9698a72b7e1aad6e40a6567d3c06d355', 5, (7, 1))
  - cancelled
  - resumed
  - resumed
  - {}
  - compute-task-1650392058.1483586
  - 2022-04-19 11:14:18.165897
```

There's definitely something weird about the `processing-released` message arriving right before the `compute-task` message. I can't find an obvious reason in scheduler code why that would happen.

But let's ignore that oddity for a second. Pretend it was just a normal work-stealing request that caused the task to be cancelled.

I find it odd that if a worker is told to _compute_ a task it was previously fetching, that it'll resume the fetch:
https://github.com/dask/distributed/blob/c9dcbe7ee87be83fde1156f18e88ebe2da992c0c/distributed/worker.py#L2269-L2271

If previously we were fetching a key, but now we're being asked to compute it, it seems almost certain that the fetch is going to fail. The compute request should probably take precedence.

I imagine here that we're assuming the `gather_dep` will error out sometime in the future, and when it does, _then_ the key will go from `resumed` to `waiting`?

Also, this is coming from the https://github.com/dask/distributed/issues/6110 scenario. That's an unusual one in that the TCP connection to the stuck worker doesn't get broken, it's just unresponsive. So I'm also wondering if perhaps `gather_dep` to the stuck worker will hang forever? for 300s (seems to go much longer than that)? for 300s * some retries? Basically, could it be that this isn't _quite_ a deadlock, but a very, very, very long wait for a dependency fetch that might never return until the other worker properly dies? If we don't have any explicit timeouts on `gather_dep` already, maybe we should.
(All that said, I still think the proper fix would be to not have `transition_cancelled_waiting` try to resume the fetch, but instead go down the compute path. The timeout might be something in addition.)

cc @fjetter

**Comment by fjetter** [comment]
This behaviour should be captured by [`test_flight_to_executing_via_cancelled_resumed`](https://github.com/dask/distributed/blob/c9dcbe7ee87be83fde1156f18e88ebe2da992c0c/distributed/tests/test_cancelled_state.py#L143-L172)


xref https://github.com/dask/distributed/pull/6112 which introduced a subtle change to the behaviour connected to this edge case

**Comment by fjetter** [comment]
This looks like the comm is never broken, i.e. the worker never detects that the remote is dead. Are there some kind of network proxies involved that would keep the TCP connection alive?

**Comment by fjetter** [comment]
Are you using the default tornado TCP backend? I've seen such a condition already with a user who implemented their own comm backend which did not properly close a comm if the peer died.

**Comment by fjetter** [comment]
> I think that's the core design problem, and what I meant by "I find it odd". Yes, the code is working as intended, because it's the best we could do with the gather_dep system. It's just annoying that since we don't have a handle on the gather, our only choice is to wait for it to finish. Even if we just had a dict[str, asyncio.Task] mapping keys to gather calls (though more complex than that since multiple keys come from a single gather), being able to call cancel() on the asyncio task would make this much easier.

Agreed. I thought about this for a while and considered introducing a way to cancel requests a bunch of times but it is not trivial. At the same time, we have a similar problem with execute and I chose to go for the "boring" approach of just waiting it out. I'm happy to revisit this at one point in time.

> This is exactly what we're seeing on the scheduler side (see https://github.com/dask/distributed/issues/6110#issuecomment-1099539228 and https://github.com/dask/distributed/issues/6148), so it seems reasonable to assume the same thing is happening in worker<->worker comms.

My knowledge on low level linux is very limited but to my best knowledge, the TCP connection would never be broken up if the kernel is still alive and kicking but the python process is frozen for whatever reason (let it be GIL, an actual process suspension or any other linux wizardry) because the kernel itself will handle TCP keepalive probes (see below)

> On the scheduler, the comm isn't even timing out after 300s (and the deadlock doesn't resolve itself after 300s), so I'm not sure that that 300s limit (where does it even come from? are you thinking of [distributed.comms.timeouts](https://github.com/dask/distributed/blob/c9dcbe7ee87be83fde1156f18e88ebe2da992c0c/distributed/distributed.yaml#L212-L214), which are both 30s by default?) is even coming into play. The TCP connection manages to stay alive and healthy, just nothing gets sent over it.


I don't know where the 300s is coming from. You mentioned this in the OP. Note, that most of the timeouts we configure is on application layer, not on network/tcp layer with the exception of what's configured in [`set_tcp_timeout`](https://github.com/dask/distributed/blob/7a564663bfda4949e5cc088a7cd66511e2d85e3f/distributed/comm/tcp.py#L65-L121). The most important configuration there is [TCP keepalive](https://tldp.org/HOWTO/TCP-Keepalive-HOWTO/overview.html) to check for dead peers. See also [linux docs](https://man7.org/linux/man-pages/man7/tcp.7.html), specifically `TCP_USER_TIMEOUT`, `TCP_KEEPCNT`, `TCP_KEEPIDLE` and `TCP_KEEPINTVL`.
This behaviour is controlled via `comm.timeouts.tcp`. 
TLDR the kernel emits keep-alive probes at regular intervals. if a remote appears to be dead and doesn't respond to keep alive probes for `distributed.comm.timeouts.tcp`s the socket is closed. That's how we currently detect dead remotes.

**Comment by fjetter** [comment]
FYI https://github.com/dask/distributed/issues/6169 but I don't think it's the same problem since for the above condition to trigger, we'd need to see a `busy-gather` log.

**Comment by fjetter** [comment]
I also floated the idea of a circuit breaker pattern once where the scheduler would be the one entity responsible for detecting dead remotes. If a worker would be flagged as dead, a signal would be broadcasted to all workers notifying them about the dead peer and they'd be instructed to abort all comms to that address. For instance, coupled with a mandatory worker-ttl, this would remove our dependency on TCP keep alives. There are obviously pros and cons

**Comment by gjoseph92** [comment]
> At the same time, we have a similar problem with execute and I chose to go for the "boring" approach of just waiting it out.

`execute` is a bit different, because we can't actually cancel it (currently, see https://github.com/dask/distributed/issues/4694). So we don't have any choice but to wait it out.

With `gather_dep`, we should be able to cancel it quite nicely thanks to asyncio (if we actually use async well https://github.com/dask/distributed/issues/6047).

> I thought about this for a while and considered introducing a way to cancel requests a bunch of times but it is not trivial.

Maybe we're thinking about different things. But what if instead of an interface like `request_gather(key: str)`, `release_gather_request(key: str)`, where we have to cooperate between multiple tasks that depend on fetching the same key, and could almost end up in a use-after-free-style situation, we instead separate the single physical network request from each task's need for the key, using a reference-counting-style approach:

```python
def request_gather(key: str, who_has: list[str]) -> GatherRequest:
    """
    Request that a key be fetched from peer workers.

    Multiple requests for the same key will return different `GatherRequest` objects,
    though each will be backed by the same underlying request.
    When the underlying request completes, all `GatherRequest`s will complete.
    If the underlying request is cancelled (because no `GatherRequest`s need it anymore,
    or it is explicitly cancelled), all `GatherRequest`s will be cancelled.
    """

class GatherRequest:
    key: str
    def release(self) -> None:
        """
        Note that this gather request is no longer needed. Idempotent.
        If this is the last `GatherRequest` needing the key, the underlying
        request is cancelled.
        """

    def cancel(self) -> None:
        """
        Explicitly cancel the underlying request for all `GatherRequest`s needing this key. Idempotent.
        Use when we know the key should not be fetched anymore (the task has instead been stolen
        to this worker to compute, for example).
        """

    def __del__(self):
        self.release()

    def __await__(self) -> Awaitable[None]
        "Wait until `key` is available in `worker.data`. If underlying request is cancelled, raises CancelledError."
```

This might be easier to reason about and test.

**Comment by gjoseph92** [comment]
> I also floated the idea of a circuit breaker pattern once where the scheduler would be the one entity responsible for detecting dead remotes. If a worker would be flagged as dead, a signal would be broadcasted to all workers notifying them about the dead peer and they'd be instructed to abort all comms to that address.

I think this would be a good idea. It's otherwise hard for a worker to know if the peer it's talking to is unresponsive. Setting a timeout on the `gather_dep` is not ideal. Getting low-enough level in the networking stack to say something like "if no new bytes have arrived over this socket in N seconds, the comm is dead" seems a bit complicated. The scheduler is already heartbeating with all workers, so it has a clear signal for this.

**Comment by fjetter** [comment]
This issue has become quite stale and the code changed drastically since the report. I'm inclined to close this issue, @gjoseph92 objections?

**Comment by gjoseph92** [comment]
Since there's no reproducer, sure.
